File: Data Structures & Algorithms/partition-equal-subset-sum/submission-1.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        total = sum(nums)
        if total % 2 != 0:
            return False
        dp = set()
        dp.add(0)
        target = total // 2
        for num in nums:
            for val in dp.copy():
                dp.add(val + num)
                if target in dp:
                    return True
        return False

File: Data Structures & Algorithms/combination-target-sum/submission-1.cpp
class Solution {
private:
    int n;
    vector<vector<int>> result;
    void backtracking(vector<int>& nums, vector<int>& comb, const int& target, int i, int sum)
    {
        if(sum == target)
            result.push_back(comb);
        else
        {
            if(i == n || sum > target)
                return;

            comb.push_back(nums[i]);
            backtracking(nums, comb, target, i, sum + nums[i]);
            comb.pop_back();
            backtracking(nums, comb, target, i+1, sum);
        }
    }
public:
    vector<vector<int>> combinationSum(vector<int>& nums, int target) 
    {
        n = nums.size();
        vector<int> comb;
        backtracking(nums, comb, target, 0, 0);
        return result;
    }
};
